### PR TITLE
Fix autoray due to newly released version

### PR DESCRIPTION
**Pull request by albi3ro**

### CODE REVIEW

File: pennylane/ops/qubit/parametric_ops.py [hunk 1]
```diff
@@ -2808,6 +2808,8 @@ class U2(Operation):
     parameter_frequencies = [(1,), (1,)]
 
     def __init__(self, phi, delta, wires, do_queue=True, id=None):
+        print(phi)
+        print(delta)
```
Comment (albi3ro) [inline pennylane/ops/qubit/parametric_ops.py:2812]:
```suggestion
```

File: doc/releases/changelog-dev.md [hunk 2]
```diff
@@ -36,6 +36,9 @@
 
 <h3>Bug fixes</h3>
 
+* Registers `math.ndim` and `math.shape` for built-ins and autoray to accomdate Autoray 0.6.1.
```
Comment (albi3ro) [inline doc/releases/changelog-dev.md:39]:
```suggestion
* Registers `math.ndim` and `math.shape` for built-ins and autograd to accomdate Autoray 0.6.1.
```

File: doc/releases/changelog-dev.md [hunk 3]
```diff
@@ -36,6 +36,9 @@
 
 <h3>Bug fixes</h3>
 
+* Registers `math.ndim` and `math.shape` for built-ins and autograd to accomdate Autoray 0.6.1.
```
Comment (rmoyard) [inline doc/releases/changelog-dev.md:39]:
```suggestion
* Registers `math.ndim` and `math.shape` for built-ins and autograd to accomodate Autoray 0.6.1.
```

File: pennylane/math/single_dispatch.py [hunk 4]
```diff
@@ -103,6 +108,8 @@ def _cond(pred, true_fn, false_fn, args):
 # qml.numpy rather than autograd.numpy, to take into account our autograd modification.
 ar.autoray._MODULE_ALIASES["autograd"] = "pennylane.numpy"
 
+ar.register_function("autograd", "ndim", lambda x: _i("qml").numpy.ndim(x))
```
Comment (rmoyard) [inline pennylane/math/single_dispatch.py:111]: I think that should work with "autograd" instead of "qml".

File: tests/ops/qubit/test_sparse.py [hunk 5]
```diff
@@ -120,7 +120,7 @@ def circuit(param):
             match="SparseHamiltonian observable must be"
             " used with the parameter-shift differentiation method",
         ):
-            qml.grad(circuit, argnum=0)([0.5])
+            qml.grad(circuit, argnum=0)(0.5)
```
Comment (rmoyard) [inline tests/ops/qubit/test_sparse.py:123]: How is it related? I see that it is batching where we do not need it but it should run and work

**Review by rmoyard — COMMENTED** [review]
Thanks for the quick fix, some questions

File: tests/ops/qubit/test_sparse.py (hunk 5 above)
Comment (albi3ro) [reply tests/ops/qubit/test_sparse.py:123]: was fixed by a later change.

**Comment by timmysilv** [comment]
The autoray [commit that caused this](https://github.com/jcmgray/autoray/commit/147c842439b889bf3a388b3018a48dd51c509604) has nearly 1200 lines added and removed. I'm assuming it's just a refactor, but what I'm getting at is - are we sure this is enough to be compatible with the latest autoray? It might be better (especially for a bugfix release) to just pin to the last working version of autoray in setup.py/requirements.txt. I'm generally hesitant with forward fixes, but our test coverage is quite extensive so I'm ok with trying this out.

File: pennylane/math/single_dispatch.py (hunk 4 above)
Comment (albi3ro) [reply pennylane/math/single_dispatch.py]: using autograd now.

**Comment by trbromley** [comment]
> The autoray [commit that caused this](https://github.com/jcmgray/autoray/commit/147c842439b889bf3a388b3018a48dd51c509604) has nearly 1200 lines added and removed. I'm assuming it's just a refactor, but what I'm getting at is - are we sure this is enough to be compatible with the latest autoray? It might be better (especially for a bugfix release) to just pin to the last working version of autoray in setup.py/requirements.txt. I'm generally hesitant with forward fixes, but our test coverage is quite extensive so I'm ok with trying this out.

Good point @timmysilv! I'm open to:
- Trying to support latest `autoray` in development and for the PL `0.30` release
- Pinning to `autoray <= 0.6.0` for the bugfix release. @albi3ro any worries about doing this?

**Review by rmoyard — APPROVED** [review]
Thanks @albi3ro it looks good to me !

File: pennylane/math/single_dispatch.py [hunk 6]
```diff
@@ -33,6 +33,11 @@ def _i(name):
     return import_module(name)
 
 
+# ------------------------------- Builtins -------------------------------- #
+
+ar.register_function("builtins", "ndim", lambda x: np.ndim(np.array(x)))
+ar.register_function("builtins", "shape", lambda x: np.array(x).shape)
```
Comment (dwierichs) [inline pennylane/math/single_dispatch.py:39]: We could use `np.array(x).ndim` for `ndim`, similar to `shape`. Is there a reason to prefer one over the other (I'd guess attribute lookup should be faster?)

**Review by dwierichs — APPROVED** [review]
Just one tiny comment, otherwise looks good to me! Thanks @albi3ro

**Review by timmysilv — COMMENTED** [review]
how will this release work? should this PR be against v0.29 instead of master?

File: pennylane/math/single_dispatch.py (hunk 6 above)
Comment (albi3ro) [reply pennylane/math/single_dispatch.py:39]: It works for now.

**Pull request merged**